tests/cpp/test_inlining.cpp

```diff
@@ -0,0 +1,325 @@
+  auto cg_outputs = fe.runFusion({input});
+
+  testValidate(&fusion, cg_outputs, {input}, __LINE__, __FILE__);
+}
+
+TEST_F(InliningTest, MatchedLeafPosWithoutReplayBroadcast) {
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  auto tv0 = makeConcreteTensor({2, 3, 4});
+  fusion.addInput(tv0);
+  auto tv1 = broadcast(tv0, {false, true, false, true, false, true});
+  auto tv2 = sin(tv1);
+  fusion.addOutput(tv2);
+
+  for (auto tv : {tv1, tv2}) {
+    tv->merge(0);
+    tv->merge(1);
+    tv->merge(2);
+  }
+
+  EXPECT_EQ(
+      TransformReplay::getMatchedLeafPosWithoutReplayPasC(tv0, tv1, 3), 3);
+  EXPECT_EQ(
+      TransformReplay::getMatchedLeafPosWithoutReplayCasP(tv1, tv0, 3), 3);
+  EXPECT_EQ(
+      TransformReplay::getMatchedLeafPosWithoutReplayPasC(tv1, tv2, 3), 3);
+  EXPECT_EQ(
+      TransformReplay::getMatchedLeafPosWithoutReplayCasP(tv2, tv1, 3), 3);
+}
+
+// Test isAllowedID with setLoopDomain
+TEST_F(InliningTest, IsAllowedID) {
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  std::vector<int64_t> shape1({3, 4, 5});
+  std::vector<int64_t> shape2({3, 4 * 5});
+
+  // Inner normalization pattern
+
+  // [i0, i1, i2]
+  auto tv0 = makeConcreteTensor(shape1);
+  fusion.addInput(tv0);
+
+  // [i0, i1*i2]
+  auto tv1 = reshape(tv0, shape1, shape2);
+  // [i0, r1*r2]
+  auto tv2 = sum(tv1, {1});
+  // [i0]
+  auto tv3 = add(tv2, fusion.oneVal());
+  // [i0, b3]
+  auto tv4 = broadcast(tv3, {false, true});
+  // [i0, i1*i2]
+  auto tv5 = sub(tv1, tv4);
+  // [i0, i1, i2]
+  auto tv6 = reshape(tv5, shape2, shape1);
+  fusion.addOutput(tv6);
+
+  MaxPosCalculator calc;
+
+  auto isAllowedID = [&calc](TensorView* tv, int64_t axis) -> bool {
+    return calc.isAllowedID(
+        tv->getLoopDomain().at(axis),
+        tv,
+        /*best_effort=*/true,
+        /*allow_reduction=*/false,
+        /*allow_vectorize=*/false,
+        /*allow_unmappable=*/false);
+  };
+
+  // First, check isAllowedID without manipulating loop domains. tv1
+  // is a persistent tensor. The inner domain needs to be persistent.
+  EXPECT_TRUE(isAllowedID(tv1, 0)) << tv1->getLoopDomain().at(0)->toString();
+  // The inner domain should not be allowed since it's a persistent domain
+  EXPECT_FALSE(isAllowedID(tv1, 1)) << tv1->getLoopDomain().at(1)->toString();
+
+  // Set [i0, i1, i2] as the loop domain of each of tensors. i1 and i2
+  // become pesistent domains.
+
+  // tv1
+  {
+    tv1->setLoopDomain(tv1->getRootDomain());
+    EXPECT_TRUE(isAllowedID(tv1, 0)) << tv1->getLoopDomain().at(0)->toString();
+    // Persistent domain
+    EXPECT_FALSE(isAllowedID(tv1, 1)) << tv1->getLoopDomain().at(1)->toString();
+    // Persistent domain
+    EXPECT_FALSE(isAllowedID(tv1, 2)) << tv1->getLoopDomain().at(2)->toString();
+  }
+
+  // tv2
+  {
+    std::vector<IterDomain*> loop_domain{
+        tv2->getLogicalDomain().at(0),
+        tv1->getLoopDomain().at(1)->cloneWithoutRFactor(),
+        tv1->getLoopDomain().at(2)->cloneWithoutRFactor()};
```

Comment:
Do we need to change the `IterType` to `Reduction`?

Reply:
That's a good point. Yes, I think so. We should actually probably do some check when doing `setLoopDomain`.

Reply:
Changed. Also #2974